### PR TITLE
Circuit for opcode DIV & MOD

### PR DESCRIPTION
**Pull request by icemelon**

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution/div.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            STOP
+        };
+        let block = witness::build_block_from_trace_code_at_start(&bytecode);
+        assert_eq!(run_test_circuit_incomplete_fixed_table(block), Ok(()));
+    }
+
+    #[test]
+    fn div_gadget_simple() {
+        test_ok(OpcodeId::DIV, 0xFF.into(), 0x4.into());
+    }
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/div.rs:103]: I suggest add more tests for arithmetic opcodes.  divided with rem, no rem, big int(at least more than one u8?), negative

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/execution/div.rs:103]: besides, we may consider using fuzz test later.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 2]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+        //radix_constant_64 == 2^64
+        //radix_constant_128 == 2^128
+        let radix_constant_64 = pow_of_two_expr(64);
+        let radix_constant_128 = pow_of_two_expr(128);
+        cb.require_equal(
+            "product(quotient, divisor)_lo + remainders_lo == dividends_lo + radix_lo ⋅ 2^128",
+            cur_v0.clone() * radix_constant_128,
+            t0.expr() + t1.expr() * radix_constant_64.clone()
+                + remainder_lo
+                - (dividend_limbs[0].clone()
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:891]: change it to `cb.require_equal("...", t0 + t1 * radix_constant_64 + remainders_lo, dividends_lo + cur_v0 * radix_constant_128)` to match the description.

btw. what's `radix_lo * 2^128` in the description? Isn't `radix_lo` equal to `2^128`?

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 3]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+                    + dividend_limbs[1].clone() * radix_constant_64.clone()),
+        );
+        cb.require_zero(
+            "product(quotient, divisor)_high + remainders_high == dividends_high",
+            cur_v0 + t2.expr() + t3.expr() * radix_constant_64.clone()
+                + remainder_hi
+                - (dividend_limbs[2].clone() + dividend_limbs[3].clone() * radix_constant_64)
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:898]: ditto

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 4]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+        let constant_64 = BigUint::from(1u128 << 64);
+        let constant_128 = constant_64.clone() * constant_64.clone();
+
+        let a_limbs = divisor.to_u64_digits();
+        let b_limbs = quotient.to_u64_digits();
+        let c_limbs = dividend.to_u64_digits();
+        let d_limbs = remainder.to_u64_digits();
+        let mut t_digits = vec![];
+
+        //a->divisor
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:967]: I think we just don't do this renaming.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 5]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+        self.lt_lo.assign(
+            region,
+            offset,
+            {
+                if remainder8s.len() <= 16 {
+                    from_bytes::value(&remainder8s[0..remainder8s.len()])
+                } else {
+                    from_bytes::value(&remainder8s[0..16])
+                }
+            },
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1033]:
```suggestion
            from_bytes::value(&remainder8s[0..min(remainder8s.len(), 16)]),
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 6]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+            {
+                if divisor8s.len() <= 16 {
+                    from_bytes::value(&divisor8s[0..divisor8s.len()])
+                } else {
+                    from_bytes::value(&divisor8s[0..16])
+                }
+            },
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1040]: ditto

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 7]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+        self.comparison_hi.assign(
+            region,
+            offset,
+            {
+                if remainder8s.len() <= 16 {
+                    F::zero()
+                } else {
+                    from_bytes::value(&remainder8s[16..remainder8s.len()])
+                }
+            },
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1051]:
```suggestion
                if remainder8s.len() <= 16 {
                    F::zero()
                } else {
                    from_bytes::value(&remainder8s[16..remainder8s.len()])
                },
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 8]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+            {
+                if divisor8s.len() <= 16 {
+                    F::zero()
+                } else {
+                    from_bytes::value(&divisor8s[16..divisor8s.len()])
+                }
+            },
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:1058]: ditto

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 9]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        dividend: util::Word<F>,
+        divisor: util::Word<F>,
+    ) -> Self {
+        let quotient = cb.query_word();
+        let remainder = cb.query_word();
+        let v0 = cb.query_bytes();
+
+        let mut dividend_limbs = vec![];
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:840]: we only need `dividend_hi` and `dividend_lo`, similar to remainder

Comment (icemelon) [reply zkevm-circuits/src/evm_circuit/util/math_gadget.rs:840]: Could you also fix this for `MulWordsGadget`?

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 10]
```diff
@@ -816,3 +816,248 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
...
+                    BigUint::from(b_limbs[b_idx])
+                } else {
+                    BigUint::from(0u128)
+                };
+                rhs_sum = rhs_sum.clone() + tmp_a * tmp_b;
+            }
+            t_digits.push(rhs_sum);
+        }
+
+        let mut c_now = vec![];
```
Comment (icemelon) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:991]: we only need `c_hi, c_lo, d_hi, d_lo`.

**Comment by noel2004** [comment]
It seems current the div gadget can not constraint quotient correctly in some cases. Consider:

1. dividend is zero
2. divisor is only non-zero in its highest 64-bit limb

Now under current circuit, one can set quotient to any non-zero value on its highest 64-bit limb rather than zero.

I have made a patch for providing this case.

Apply following patch and run `cargo test div_gadget_malform -- --nocapture` you can see the malicious assignation on quotient has been triggered while the test still pass.

``` diff

diff --git a/zkevm-circuits/src/evm_circuit/execution/div.rs b/zkevm-circuits/src/evm_circuit/execution/div.rs
index 1f53583..61cc3f8 100644
--- a/zkevm-circuits/src/evm_circuit/execution/div.rs
+++ b/zkevm-circuits/src/evm_circuit/execution/div.rs
@@ -108,6 +108,14 @@ mod test {
         );
     }
 
+    #[test]
+    fn div_gadget_malform() {
+        let mut divend = [0u8; 32];
+        divend[0] = 64u8;
+
+        test_ok(OpcodeId::DIV, 0x0.into(), Word::from_big_endian(&divend));
+    }
+
     #[test]
     fn div_gadget_rand() {
         let dividend = rand_word();
diff --git a/zkevm-circuits/src/evm_circuit/util/math_gadget.rs b/zkevm-circuits/src/evm_circuit/util/math_gadget.rs
index 0c0d4bc..29a200d 100644
--- a/zkevm-circuits/src/evm_circuit/util/math_gadget.rs
+++ b/zkevm-circuits/src/evm_circuit/util/math_gadget.rs
@@ -950,14 +950,24 @@ impl<F: FieldExt> DivWordsGadget<F> {
         wremainder: &Word,
     ) -> Result<(), Error> {
         use num::BigUint;
+        let wquotient_mal = 
+        if Word::zero().eq(wdividend) && wdivisor.trailing_zeros() > 192{
+            let mut bigw = [0u8; 32];
+            bigw[0] = 32u8;
+            Word::from_big_endian(&bigw)
+        } else {
+            wquotient.clone()
+        };
 
         let dividend = BigUint::from_bytes_le(&wdividend.to_le_bytes());
         let divisor = BigUint::from_bytes_le(&wdivisor.to_le_bytes());
-        let quotient = BigUint::from_bytes_le(&wquotient.to_le_bytes());
+        let quotient = BigUint::from_bytes_le(&wquotient_mal.to_le_bytes());
         let remainder = BigUint::from_bytes_le(&wremainder.to_le_bytes());
         let constant_64 = BigUint::from(1u128 << 64);
         let constant_128 = constant_64.clone() * constant_64.clone();
 
+        println!("malform quotient {} vs {} ", wquotient, quotient);
+
         let a_limbs = divisor.to_u64_digits();
         let b_limbs = quotient.to_u64_digits();
         let c_limbs = dividend.to_u64_digits();

```

**Comment by icemelon** [comment]
Need to close this PR and open a new one due to some bug in github

**Pull request closed without merging**